File: libs/formgen/src/lib/types/renderable-form.ts
export type FieldType =
  | 'text' | 'email' | 'password' | 'number' | 'tel' | 'url'
  | 'textarea' | 'select' | 'checkbox' | 'radio'
  | 'date' | 'slider' | 'toggle';

export interface FieldOption {
  label: string;
  value: string | number;
}

export interface ValidatorConfig {
  type: 'required' | 'minLength' | 'maxLength' | 'min' | 'max' | 'pattern' | 'email';
  value?: unknown;
  message?: string;
}

export interface RenderableLayout {
  columns: 1 | 2 | 3 | 4;
  gap: string;
  breakpoints: {
    sm: { columns: number };
    md: { columns: number };
  };
}

export interface RenderableField {
  key: string;
  type: FieldType;
  label: string;
  placeholder: string;
  hint: string;
  default_value?: unknown;
  validators: ValidatorConfig[];
  options: FieldOption[];
  col_span: number;
  order: number;
  css_class: string;
}

export interface RenderableAction {
  type: 'submit' | 'reset' | 'custom';
  label: string;
  color: 'primary' | 'accent' | 'warn';
  css_class: string;
}

export interface RenderableForm {
  id: string;
  name: string;
  layout: RenderableLayout;
  fields: RenderableField[];
  actions: RenderableAction[];
  css_overrides?: string;
}

File: apps/api/shared/renderable_form.py
from __future__ import annotations
from typing import Literal, Optional, Any
from pydantic import BaseModel


FieldType = Literal[
    "text", "email", "password", "number", "tel", "url",
    "textarea", "select", "checkbox", "radio",
    "date", "slider", "toggle",
]


class FieldOption(BaseModel):
    label: str
    value: str | int


class ValidatorConfig(BaseModel):
    type: Literal["required", "minLength", "maxLength", "min", "max", "pattern", "email"]
    value: Optional[Any] = None
    message: Optional[str] = None


class RenderableLayout(BaseModel):
    columns: Literal[1, 2, 3, 4]
    gap: str
    breakpoints: dict[str, dict[str, int]]


class RenderableField(BaseModel):
    key: str
    type: FieldType
    label: str
    placeholder: str
    hint: str
    default_value: Optional[Any] = None
    validators: list[ValidatorConfig]
    options: list[FieldOption]
    col_span: int
    order: int
    css_class: str


class RenderableAction(BaseModel):
    type: Literal["submit", "reset", "custom"]
    label: str
    color: Literal["primary", "accent", "warn"]
    css_class: str


class RenderableForm(BaseModel):
    id: str
    name: str
    layout: RenderableLayout
    fields: list[RenderableField]
    actions: list[RenderableAction]
    css_overrides: Optional[str] = None
